### PR TITLE
[FEATURE] First version of SpectraST search adapter util

### PR DESCRIPTION
**Pull request by lukaszimmermann**
Advanced options of SEARCH mode are not yet supported.

### CODE REVIEW

**Comment by timosachsenberg** [comment]
@grosenberger @hroest I think this PR might me interesting for the SWATH workflow. Could you or some of your colleagues assist Lukas with his first PR?

**Comment by grosenberger** [comment]
I think it is great that you developed a SpectraST adapter. This tool provides many different workflows that further extend the capabilities of OpenMS.

As you have probably seen, SpectraST provides a huge number of options and different modes of operation. I think your approach to focus on individual workflows is the best start. But would it also make sense to support [SpectraST Params files](http://tools.proteomecenter.org/wiki/index.php?title=Software:SpectraST#SpectraST_Parameter_Files) that would enable to use more complex sets of parameters with the generic adapter?

We have also already implemented a limited interface of SpectraST MRM files with OpenMS, that we use to directly convert spectral libraries to TraML. Usually, we start with a database search of DDA or DIA-Umpire pseudo spectra using the TPP and process the pepXML using PeptideProphet, iProphet and MAYU. Then we import these files in SpectraST and generate a consensus spectral library. The whole process is documented in this recent [protocol paper](http://www.nature.com/nprot/journal/v10/n3/abs/nprot.2015.015.html). Instead of using spectrast2tsv.py from msproteomicstools, we now can directly export MRM files from SpectraST (using the flag `-cM`). These can be used by ConvertTSVToTraML to generate a TraML file that can be further filtered by OpenSwathAssayGenerator. This would of course be great to consider for the create mode of SpectraST.

**Comment by lkszmn** [comment]
Thank you very much for the valuable information and the reference to the protocol. I am going to read it and see what I can do. I think supporting SpectraST params files would definitely make sense.

**Comment by hroest** [comment]
@grosenberger does this mean that in the future we could do this directly in TOPPAS / KNIME through this adaptor? Assuming we can give the correct param file?

**Comment by grosenberger** [comment]
Yes, I think this should be possible. We usually do not change the default parameters except I/O, so a full generic workflow should be possible within OpenMS.

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+      // Sequence database file
+      registerInputFile_(TOPPSpectraSTSearchAdapter::param_sequence_database_file, "<sequencedb_file>.fasta", "", "Specify a sequence database file",false,false);
+      setValidFormats_(TOPPSpectraSTSearchAdapter::param_sequence_database_file,ListUtils::create<String>("fasta"),false);
+
+      registerStringOption_(TOPPSpectraSTSearchAdapter::param_sequence_database_type, "<sequencedb_type>", "", "Specify type of sequence database",false, false);
+      setValidStrings_(TOPPSpectraSTSearchAdapter::param_sequence_database_type,ListUtils::create<String>("DNA,AA"));
+
+      // Search file
+      registerInputFile_(TOPPSpectraSTSearchAdapter::param_search_file, "<search_file>", "", "Only search a subset of the query spectra in the search file", false, false);
```
Comment (lkszmn) [inline src/utils/SpectraSTSearchAdapter.cpp]: The file contains a list of spectra names, one per line. Is there a convention for it? Otherwise I'll just use txt and dat.

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+      {
+        LOG_ERROR << "ERROR: Input filename does not agree in format: "
+                       << input_file << " is not " << inputFormat << endl;
+        return ILLEGAL_PARAMETERS;
+      }
+      arguments << input_file.toQString();
+     }
+
+
+     // Echoing the command
```
Comment (lkszmn) [inline src/utils/SpectraSTSearchAdapter.cpp]: Yes. It would be.

File: src/utils/SpectraSTSearchAdapter.cpp (hunk 1 above)
Comment (jpfeuffer) [inline src/utils/SpectraSTSearchAdapter.cpp]: Yes, that would be fine I think.

**Comment by jpfeuffer** [comment]
Hey @lukaszimmermann , what is the status of this PR?
There seem to be interested users. Is the search functionality ready? Do you plan to extend it further in this PR?

**Comment by lukaszimmermann** [comment]
The non-advanced parameters of the SpectraST Search Mode are implemented. In principle, the util can be used for searching, but one should compare the resulting call to spectrast from the util with the expected one (I tested several combinations, but just to make sure).

I'd be happy to add advanced options to the SpectraST Search Adapter on request. I have started to work on a similar adapter for the Create mode.

**Comment by jpfeuffer** [comment]
Sounds good. I also like the approach of splitting the adapter into Search and Create adapters.
I would say that if you all think that the searching capabilities in the current state of this PR are sufficient, that it can be merged as the UTIL that it is.

In the future we can have updates that make the tool a TOPP tool e.g. by adding a tool test like you already proposed (e.g. taking the result from the command line SpectraST and comparing the result of the adapter to it)
Or another PR for the Create functionality or param file support.

**Comment by jpfeuffer** [comment]
I will restart Travis.

**Comment by hendrikweisser** [comment]
> In the future we can have updates that make the tool a TOPP tool e.g. by adding a tool test like you already proposed (e.g. taking the result from the command line SpectraST and comparing the result of the adapter to it)

If you already have test data, it might be useful to add an automated test already now. It's not required for UTILs, but some do have tests.

**Comment by lukaszimmermann** [comment]
I am going to write tests that compare the generated with the expected call to the spectrast executable.

**Comment by hroest** [comment]
@lukaszimmermann maybe add the test into a new PR as we may need to change travis and have it download the spectrast binary -- otherwise of course the test fails :-) 

I would be ok with the PR as it is without a test as adding one would require some changes to our infrastructure, including cdash and travis

**Comment by lukaszimmermann** [comment]
Ok. I am fine with that. I wil add the tests in  a later PR

File: share/OpenMS/SCRIPTS/Plot_trafoXML.R [hunk 3]
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env Rscript
+
+library(XML)
```
Comment (timosachsenberg) [inline share/OpenMS/SCRIPTS/Plot_trafoXML.R]: BTW why is this file here?

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+using namespace std;
+
+//-------------------------------------------------------------
+// Doxygen docu
+//-------------------------------------------------------------
+
+/**
+    @page UTILS_SpectraSTSearchAdapter
+
+    @brief This tool provides an interface to the 'SEARCH' mode of the SpectraST program. All parameters rf
```
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: rf?

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+  // it gets automatically called on tool execution
+  void registerOptionsAndFlags_()
+  {
+      StringList empty;
+
+      // Handle executable
+      registerInputFile_(TOPPSpectraSTSearchAdapter::param_executable, "<path>", "spectrast", "Path to the SpectraST executable to use; may be empty if the executable is globally available.", true, false, ListUtils::create<String>("skipexists"));
+
+      // register spectra input files
+      registerInputFileList_(TOPPSpectraSTSearchAdapter::param_spectra_files, "<SearchFileName1> [ <SearchFileName2> ... <SearchFileNameN> ]", empty, "Name(s) of file containing unknown spectra to be searched.", true, false);
```
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: maybe just "File names(s) of spectra to be searched"

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+      setValidFormats_(TOPPSpectraSTSearchAdapter::param_spectra_files, ListUtils::create<String>(TOPPSpectraSTSearchAdapter::param_spectra_files_formats), false);
+
+      // register Output files
+      registerOutputFileList_(TOPPSpectraSTSearchAdapter::param_output_files,"<OutputFile1> [ <OutputFileName2> ... <OutputFileNameN> ]", empty, "Files where the output should be written to. Make sure to specify one output file for each input file.",true,false);
```
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: "Output files. Make sure to specify one output file for each input file."

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+      setValidFormats_(TOPPSpectraSTSearchAdapter::param_output_files, TOPPSpectraSTSearchAdapter::param_output_file_formats, false);
+
+
+      // Require library file to be searched  // TODO Check the existence of the .spidx file
+      registerInputFile_(TOPPSpectraSTSearchAdapter::param_library_file, "<lib_file>.splib", "", "Specify library file.", true, false);
+      setValidFormats_(TOPPSpectraSTSearchAdapter::param_library_file,ListUtils::create<String>("splib"),false);
+
+
+      // Sequence database file
+      registerInputFile_(TOPPSpectraSTSearchAdapter::param_sequence_database_file, "<sequencedb_file>.fasta", "", "Specify a sequence database file",false,false);
```
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: maybe "The sequence database."

Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: spaces after comma

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+      setValidFormats_(TOPPSpectraSTSearchAdapter::param_sequence_database_file, ListUtils::create<String>("fasta"),false);
```
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: space after comma (also multiple times below)

Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: hint: no search and replace because e.g. "," in StringList may not be separated by space

File: src/utils/SpectraSTSearchAdapter.cpp (hunk 1 above)
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: Does anyone know if this option is really used? I would make it at least advanced

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+      registerInputFile_(TOPPSpectraSTSearchAdapter::param_search_file, "<search_file>", "", "Only search a subset of the query spectra in the search file", false, false);
+      setValidFormats_(TOPPSpectraSTSearchAdapter::param_search_file, ListUtils::create<String>("txt,dat"), false);
+
+      // params file
+      registerInputFile_(TOPPSpectraSTSearchAdapter::param_params_file, "<params_file>", "", "Read search options from file. All options set in the file will be overridden by command-line options, if specified.",false,false);
+      setValidFormats_(TOPPSpectraSTSearchAdapter::param_params_file, ListUtils::create<String>("params"), false);
+
+
+      // Precursor m/z tolerance
+      registerDoubleOption_(TOPPSpectraSTSearchAdapter::param_precursor_mz_tolerance, "<precursor_mz_tolerance>", 10, "m/z tolerance within which candidate entries are compared to the query", false, false);
```
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: are this m/z (aka Thomson or Dalton) or ppm?

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+
+      //Whether to use isotope average instead of monoisotopic mass
+      registerFlag_(TOPPSpectraSTSearchAdapter::param_use_isotopically_averaged_mass, "Use isotopically averaged mass instead of monoisotopic mass", false);
+
+      // Whether to use all charge states
+      registerFlag_(TOPPSpectraSTSearchAdapter::param_use_all_charge_states, "Search library spectra of all charge states, i.e., ignore specified charge state (if any) of the query spectrum",false);
```
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: advanced?

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+      // params file
+      registerInputFile_(TOPPSpectraSTSearchAdapter::param_params_file, "<params_file>", "", "Read search options from file. All options set in the file will be overridden by command-line options, if specified.",false,false);
+      setValidFormats_(TOPPSpectraSTSearchAdapter::param_params_file, ListUtils::create<String>("params"), false);
+
+
+      // Precursor m/z tolerance
+      registerDoubleOption_(TOPPSpectraSTSearchAdapter::param_precursor_mz_tolerance, "<precursor_mz_tolerance>", 10, "m/z tolerance within which candidate entries are compared to the query", false, false);
+
+      //Whether to use isotope average instead of monoisotopic mass
+      registerFlag_(TOPPSpectraSTSearchAdapter::param_use_isotopically_averaged_mass, "Use isotopically averaged mass instead of monoisotopic mass", false);
```
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: advanced?

File: src/utils/SpectraSTSearchAdapter.cpp [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+
+      // Whether to use all charge states
+      registerFlag_(TOPPSpectraSTSearchAdapter::param_use_all_charge_states, "Search library spectra of all charge states, i.e., ignore specified charge state (if any) of the query spectrum",false);
+
+      // User defined modifications file
+      registerInputFile_(TOPPSpectraSTSearchAdapter::param_user_mod_file,"<user_mod_file>","","Specify name of user-defined modifications file. Default is \"spectrast.usermods\".",false,false);
```
Comment (timosachsenberg) [inline src/utils/SpectraSTSearchAdapter.cpp]: advanced?

**Comment by lukaszimmermann** [comment]
Thanks for reviewing. I think I addressed all points.

**Comment by timosachsenberg** [comment]
could you first merge current develop into this branch to synchronize? I currently has a lot of changes unrelated to spectrast

**Comment by lukaszimmermann** [comment]
done

**Comment by timosachsenberg** [comment]
Thanks! looks much cleaner now.

**Comment by hroest** [comment]
PS: @timosachsenberg @jpfeuffer it would be great to think about adding SpectraST to our build infrastructure as well so that we can have automated tests for it...
